### ISSUE TITLE
Failed to compile in Debian

### ISSUE THREAD

**Issue by mashingan** [SSPkrolik/nimongo#42]
I tried compile Nimongo by importing it but it failed immediately

```Nim
import nimongo.mongo
```

Just that line, it failed with error message was:

```
/home/rahma/nim/nimcache/nimongo_auth.o: In function `nimongo_authInit000':
nimongo_auth.c:(.text+0x97): multiple definition of `nimongo_authInit000'
/home/rahma/nim/nimcache/nimongo_auth.o:nimongo_auth.c:(.text+0x97): first defined here
/home/rahma/nim/nimcache/nimongo_auth.o: In function `nimongo_authDatInit000':
nimongo_auth.c:(.text+0xd0): multiple definition of `nimongo_authDatInit000'
/home/rahma/nim/nimcache/nimongo_auth.o:nimongo_auth.c:(.text+0xd0): first defined here
/usr/lib/gcc/x86_64-linux-gnu/4.9/../../../x86_64-linux-gnu/crt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
/home/rahma/nim/nimcache/nimongo_mongo.o: In function `newMongo_KMrCR5eKpKUfDf4DP9a5i1Q':
nimongo_mongo.c:(.text+0x435): undefined reference to `init_C5TJYGhjrn3WAZC436gFrQ'
/home/rahma/nim/nimcache/nimongo_mongo.o: In function `newMongoWithURI_l1RGvNbDzWJyW54OYFA1rQ':
nimongo_mongo.c:(.text+0x536): undefined reference to `init_wc5Ol0q7KfDiv1xe661FtA'
/home/rahma/nim/nimcache/nimongo_mongo.o: In function `newAsyncMongo_nMYNEvG9apPbjHuKZBtseLQ':
nimongo_mongo.c:(.text+0xce9): undefined reference to `init_C5TJYGhjrn3WAZC436gFrQ'
/home/rahma/nim/nimcache/nimongo_mongo.o: In function `newAsyncMongoWithURI_ysEbTZ9c6ZNldeQOVtKHV9aw':
nimongo_mongo.c:(.text+0xf0c): undefined reference to `init_wc5Ol0q7KfDiv1xe661FtA'
/home/rahma/nim/nimcache/nimongo_clientbase.o: In function `init_IG9a9ajkUBDcc65sXE6JOxTQ':
nimongo_clientbase.c:(.text+0x728): undefined reference to `init_C5TJYGhjrn3WAZC436gFrQ'
collect2: error: ld returned 1 exit status
```

``$ uname -a`` returns ``Linux instance-2 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2+deb8u5 (2017-09-19) x86_64 GNU/Linux``

GCC version

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.9/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 4.9.2-10' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --with-arch-32=i586 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.9.2 (Debian 4.9.2-10)
```

Any idea what did it go wrong?
(I tried in Windows 10 with msys-2, it compiles and run)

**Comment by ba0f3** [comment]
remove your nimcache directory and try again

**Comment by mashingan** [comment]
It's same even after deleting the nimcache directory.

I tried with different version of Nim, I'm using `choosenim` to install multiple Nim versions. It failed with 0.17.0 too.
With 0.16.0, the error was different, as Nim couldn't find modules `private/utils` (I think it's Nimble's problem).

After looking a bit `nimongo/mongo.nim` and error message, I think it's method overloading failure.
Maybe this problem is specific to older GCC or not, I'll try with different version of GCC later.
Also, it was failed to compile too using clang and LLVM 3.5

**Comment by SSPkrolik** [comment]
Hi @mashingan. Unfortunately I do not have Debian currently, but I can try to check on latest Ubuntu

**Comment by dom96** [comment]
@federicoceratto maybe you could give this a try?

**Comment by FedericoCeratto** [comment]
@mashingan Are you using this version of GCC from Jessie? https://packages.debian.org/jessie/gcc
Have you tried a newer GCC? 7.2.0-14 works

**Comment by mashingan** [comment]
@FedericoCeratto Yes, I'm using the official package from Jessie. I'm trying to compile newer GCC to make it separated from system GCC. I'll report back the result.
I think it should work for newer GCC, I compiled in Windows 10 using mingw-w64 with GCC 7.2 and it worked fine.

@SSPkrolik I checked on latest Mint and it's worked. On Mint, the GCC is about version 5. It worked OK, but I never able to authenticate, it happened both in Windows and Mint.
Should I report this as another issue?

Looking the error, I think I missed some silly steps installing compiler toolchains or something like that.
I did what I usually do on Ubuntu/Mint and it worked fine on those. I think I'll look more for the cause. Any advice would help.
